src/Type/Symfony/NodeDefinitionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,33 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Symfony;
+
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\DynamicMethodReturnTypeExtension;
+use PHPStan\Type\Type;
+
+final class NodeDefinitionDynamicReturnTypeExtension implements DynamicMethodReturnTypeExtension
+{
+
+	public function getClass(): string
+	{
+		return 'Symfony\Component\Config\Definition\Builder\NodeDefinition';
+	}
+
+	public function isMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === 'end';
+	}
+
+	public function getTypeFromMethodCall(
+		MethodReflection $methodReflection,
+		MethodCall $methodCall,
+		Scope $scope
+	): Type
+	{
+		return new \PHPStan\Type\ObjectType('Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition');
```

Comment:
There's no logic at all, you can simply write a stub file (or send Symfony a PR that fixes the PHPDoc). https://phpstan.org/user-guide/stub-files

Reply:
In the description I wrote why there is no logic yet: I want to know if I can somehow read the 2nd argument of the constrctor of the TreeBuilder. That one defines the type....So when the 2nd argument is `new TreeBuilder('name', 'array')` then `getRootNode()` returns `ArrayNodeDefinition`. 

There is a list of some string inputs to *NodeDefinition types... but it might be that this cannot be read from the source code...

Reply:
1) You can solve it with generics although that would require `TreeBuilder` to be generic so that each place where it's typehinted would require to specify the type, like `TreeBuilder<array>`. Then you'd be able to read the type in the extension from `GenericObjectType`.
2) Or for local code, you can write a dynamic return type extension (`DynamicStaticMethodReturnTypeExtension`) for `TreeBuilder::__construct()` which gets invoked for `new` as well. And return a new `Type` class that extends `ObjectType` and carries the additional information, which you in turn read in `getRootNode` dynamic return type extension as well. Something like this is already done in `phpstan-doctrine` in `QueryBuilderType` (https://github.com/phpstan/phpstan-doctrine/blob/master/src/Type/Doctrine/QueryBuilder/QueryBuilderType.php).

Reply:
@ondrejmirtes Thanks for the suggestion. I went for option 2. It works 🎉  Amazing.

I decided to remove the `->end()` functionality. Let's first finish and merge this. Then I'll create a new PR with tests to work on supporting `->end()`.